base/acme/src/main/java/org/dogtagpki/acme/server/ACMEEngine.java

```diff
@@ -414,6 +423,65 @@ public void initMonitors(String filename) throws Exception {
         engineConfigSource.init(monitorsConfig);
     }
 
+    public void initRealm(String filename) throws Exception {
+
+        File realmConfigFile = new File(filename);
+
+        if (realmConfigFile.exists()) {
+            logger.info("Loading ACME realm config from " + realmConfigFile);
+            Properties props = new Properties();
+            try (FileReader reader = new FileReader(realmConfigFile)) {
+                props.load(reader);
+            }
+            realmConfig = ACMERealmConfig.fromProperties(props);
+
+        } else {
+            logger.info("Loading default ACME realm config");
+            realmConfig = new ACMERealmConfig();
+        }
+
+        logger.info("Initializing ACME realm");
+
+        String className = realmConfig.getClassName();
+        Class<ACMERealm> realmClass = (Class<ACMERealm>) Class.forName(className);
+
+        realm = realmClass.newInstance();
+        realm.setConfig(realmConfig);
+        realm.init();
+
+        ProxyRealm.registerRealm(name, new RealmBase() {
+            @Override
+            public Principal getPrincipal(String username) {
+                return null;
```

Comment:
Just curious, but why wouldn't we call `realm.getPrincipal()` here? Then it is up to the particular realm to return `null` or not, and if they're later necessarily, this code will work.

Reply:
I could be wrong, but my guess is the `getPrincipal()` and `getPassword()` are probably used to pull plain text passwords from the database so Tomcat can compare it, so I'm not sure if we want to use it. For our purposes we only need the two `authenticate()` methods, one for basic auth and another for client cert auth.

Reply:
Hmmm... You're right, it appears to be unused in Tomcat.